shared/libpng-util/libpng-util.cpp

```diff
@@ -0,0 +1,122 @@
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <string>
+
+#include <png.h>
+#include <stdlib.h> // for malloc/free
+
+unsigned libpng_encode32_file(const unsigned char* image, const unsigned w, const unsigned h, const char* filename) {
+  FILE *fp = fopen(filename, "wb");
+
+  png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+
+  // Write header (8 bit colour depth)
+  png_set_IHDR(png, info, w, h,
+               8, PNG_COLOR_TYPE_RGBA, PNG_INTERLACE_NONE,
+               PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
+
+  png_write_info(png, info);
+
+  for (unsigned i = 0; i < h; ++i) {
+    png_write_row(png, (png_bytep)&image[sizeof(png_byte) * 4 * w * i]);
+  }
+  png_write_end(png, NULL);
+
+  fclose(fp);
+  png_free_data(png, info, PNG_FREE_ALL, -1);
+  png_destroy_write_struct(&png, (png_infopp)NULL);
+
+  return 0;
+}
+
+unsigned libpng_decode32_file(unsigned char** out, unsigned* w, unsigned* h, const char* filename) {
+  (*w) = 0; (*h) = 0;
+  FILE *fp = fopen(filename, "rb");
+
+  png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+  png_read_info(png, info);
+
+  png_byte color_type, bit_depth;
+  (*w)       = png_get_image_width(png, info);
+  (*h)       = png_get_image_height(png, info);
+  color_type = png_get_color_type(png, info);
+  bit_depth  = png_get_bit_depth(png, info);
+
+  // Read any color_type into 8bit depth, RGBA format.
+  // See http://www.libpng.org/pub/png/libpng-manual.txt
+
+  if (bit_depth == 16)
+    png_set_strip_16(png);
+
+  if (color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_palette_to_rgb(png);
+
+  // PNG_COLOR_TYPE_GRAY_ALPHA is always 8 or 16bit depth.
+  if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
+    png_set_expand_gray_1_2_4_to_8(png);
+
+  if (png_get_valid(png, info, PNG_INFO_tRNS))
+    png_set_tRNS_to_alpha(png);
+
+  // These color_type don't have an alpha channel then fill it with 0xff.
+  if (color_type == PNG_COLOR_TYPE_RGB ||
+      color_type == PNG_COLOR_TYPE_GRAY ||
+      color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_filler(png, 0xFF, PNG_FILLER_AFTER);
+
+  if (color_type == PNG_COLOR_TYPE_GRAY ||
+      color_type == PNG_COLOR_TYPE_GRAY_ALPHA)
+    png_set_gray_to_rgb(png);
+
+  png_read_update_info(png, info);
+
+  png_bytep image;
+  size_t pitch = sizeof(png_byte) * 4 * (*w); // number of bytes in a row
+  image = (png_bytep)malloc(pitch * (*h));
```

Comment:
Why are we using `malloc`?

Reply:
Uhm, I don't know, because libpng is a C library, I didn't really mean to give the filename a `cpp` extension. However, I can change it to new easily, it doesn't really matter.

Reply:
C doesn't care how you reserved those bytes.

Reply:
Alright: 9076bb3745e0fd4f32ed6c121181a20e81e9b905

That's true, I did know this, but wasn't thinking about it. It's good that you pointed it out explicitly because it helped me realize I don't need to do the extra allocation I was doing in `loadImageData` for GameData.

Reply:
There's plenty of places left to optimize, but I think we should forego the rest for now because we really do need to get this into master. One thing that is really going to improve the situation later is having texture paging which will help us eliminate most of the remaining extra allocation.